### DISCUSSION TITLE
Spend Allocation

### DISCUSSION THREAD

**Discussion by Dhrumil-Sentry** [getsentry/sentry#44562] [Launch Feedback]
## Update

This feature is no longer available on the latest business plans

----

Our users want to use Sentry for monitoring all their apps across platforms. However, some feedback we’ve been hearing from our [largest customers](https://sentry.io/customers/) goes along the lines of - “We want to ensure that critical projects like our main user-facing app are guaranteed their  budgeted share of errors and aren’t impacted by other projects with heavy consumption.”

This problem manifests either as a project with runaway consumption going through the entire event volume for that customer causing a monitoring blackout for critical projects or a team that paid for its usage of Sentry not getting its “fair share” of monitoring coverage. Sometimes, it's both! This is typically a challenge for enterprise customers that have multiple teams and multiple business groups.

We have taken a two-pronged approach to solve this problem :

1. Launching [Spike Protection 2.0](https://github.com/getsentry/sentry/discussions/44155) to protect users from any quota-depleting spikes
2. Building Spend Allocation, a feature that allows users to allocate events to key projects

Spend Allocation gives our customers the ability to use their reserved Sentry events based on their monitoring needs by allocating volumes of their [reserved errors](https://docs.sentry.io/product/accounts/pricing/#prepaid-reserved-capacity), transactions, and attachments to specific projects.
## How does Spend Allocation work?

Let’s say Acme Corp. has 1M reserved monthly errors across 20 projects being monitored by Sentry and their main app is monitored using just two projects, A on the frontend and B on the backend.  

With Spend Allocation, Acme can specify project-level allocations for A and B for the current billing period, say 300K and 250K errors respectively. This leaves an unallocated pool of 450K errors that may be consumed by other projects at Acme. Here’s what the allocation looks like:

- Project *A* allocated 300K errors
- Project B allocated 250K errors
- Unallocated pool of 450K errors available for other projects

The allocations ensure that projects A and B get 300K and 250K errors at a minimum, while other projects would consume from the unallocated pool of 450K errors. Simply put, a project with high consumption will not impact monitoring for projects A & B because both projects have guaranteed minimum allocations.

Consider two scenarios for project A:

1. Project A exceeds its 300K error allocation
    - When A nears its allocated error budget, key project members (users with Owners and Billing roles) will be notified by email. Notifications are triggered at 80% and 100% consumption of the allocation.
    - After consuming the allocated 300K errors, A will start consuming from the 450K unallocated pool of errors. Once the 450K unallocated pool is exhausted A will start consuming the on-demand spend if available and otherwise new events will be dropped for A and other projects. Project B’s allocated errors will not be impacted and B will continue to ingest errors till it reaches its allocated 250K budget.
2. Project A does not use its complete error allocation
    - Any allocation for A lasts till the end of that month and is not carried over into the next monthly cycle and a fresh allocation is made at the start of the next month.

## Enable Spend Allocation for your Org

You must have the Owner or Billing role to enable spend allocation for your Org

1. Go to Org Settings→Spend Allocation 
![image](https://user-images.githubusercontent.com/109983869/229863875-dcc88361-1568-4229-90e4-1fe67c6a89d9.png)
2. Click on the New Allocation button to add a project allocation for your projects
![image](https://user-images.githubusercontent.com/109983869/229863926-c2fab476-c48c-4eac-97d7-304e38eaa9b5.png)
![image](https://user-images.githubusercontent.com/109983869/229863958-c2b07975-f861-47d6-a1df-931d008b73b2.png)
3. You can see the information to help you manage your allocations
    1. Unallocated Error Pool- What is available and what has been consumed
    2. Allocations made per project
4. You can adjust allocations in the current billing period if there are enough unallocated events left to honor your adjusted allocation
5. You can use Spend Allocation for reserved errors, transactions, and attachments

## We want to hear from you!

Please let us know if you have any feedback, Specifically, we’d appreciate any feedback on the following

1. Does Spend Allocation make it easier for you to ensure reserved events are saved for key projects?
2. Do you have enough visibility into how you can set allocation quantities?
4. Do you want the allocations to be made to teams instead of projects?

Feel free to comment below or drop us a note at [feedback-spend-allocation@sentry.io](mailto:feedback-spend-allocation@sentry.io)

**Comment by abey-alex** [comment]
Spend allocations either at a team level or if there was a bulk-update view for projects would be a lot easier to operate. The current flow of updating one project time is tedious if there are too many projects.

**Reply by Dhrumil-Sentry** [reply]
Hi @abey-alex thanks a lot for the feedback!

A couple of questions to understand your use case better. 

1. Why do you want to set up allocations for multiple projects? 
2. Is it to guarantee a min number of errors for each project or is it meant to protect them from spikes?
3. Would APIs to add allocations make this easier for you? 

The allocations would get copied to the next period every month so this would mainly be a 1-time setup for most customers but I definitely agree that a bulk-update view would be useful here. 

Also, We chose not to take the team-level approach because our customers often have multiple teams working on the same projects and so the allocations in such cases would have become confusing.

**Reply by abey-alex** [reply]
Thanks for the response. Mostly 2, to safeguard from spikes of one project exhausting the quota for the entire company. This works better if project ownership is not shared across teams. An API won't make a difference in the usability of the feature, at least in our case.

**Reply by Dhrumil-Sentry** [reply]
Hi @abey-alex , Thanks for that feedback. I'll add this to our backlog but it will be awhile before we make further improvements here

> Mostly 2, to safeguard from spikes of one project exhausting the quota for the entire company

Wrt to this problem you called out, I'd suggest checking out our improved Spike Protection: https://github.com/getsentry/sentry/discussions/44155

It's currently in Beta but it will be generally available in a few weeks

**Comment by Dhrumil-Sentry** [comment]
Spend Allocation is now available to all customers on a Business plan and above. Please see this [blog post](https://blog.sentry.io/2023/04/18/multiple-projects-on-sentry-new-spend-allocation-spike-protection/) to learn how Spend Allocation and improved spike protection helps you manage your quotas better

**Comment by vaibhavagrawal2912** [comment]
Do we have alert integrations for spend allocation over Slack/emails/any other? 

If any project is near to end of the budget, can we configure the mail alerts or slack alerts to specific teams/persons so that any action item can be taken before the budget is exceeded? 

Want this data to be somehow presented as metrics as well to capture the details of which projects are consuming quota on monthly basis.

**Reply by Dhrumil-Sentry** [reply]
HI @vaibhavagrawal2912 , Emails are sent to users with Owner and billing roles when a project consumes 80% and 100% of their allocations. 

It's not possible to send these alerts to specific teams/users today but this is helpful feedback for us. 

Wrt to your other ask about metrics, do you find the information in the [Stats](https://docs.sentry.io/product/stats/) page inadequate to track usage?

**Reply by vaibhavagrawal2912** [reply]
Yes, the stats page helps us in visualizing the data. Thanks

**Reply by cdignam-segment** [reply]
Is this still accurate?

> HI @vaibhavagrawal2912 , Emails are sent to users with Owner and billing roles when a project consumes 80% and 100% of their allocations.

The docs say "Team members with Manager, Billing, or Owner-level permissions will be notified via email when 80% of your Spend Allocation for each project has been consumed, and then again when 100% has been used up." https://docs.sentry.io/product/accounts/quotas/spend-allocation/#how-spend-allocation-works

**Reply by Dhrumil-Sentry** [reply]
Hi @cdignam-segment, It should still be accurate- Did it not work for you?

**Reply by cdignam-segment** [reply]
@Dhrumil-Sentry My bad, I think that is working. We had some confusing behavior because of email forwards that were setup

**Reply by vsanse** [reply]
is there a way to configure these spend allocation exhaust alerts at 80% and 100% to be sent to a slack channel or any way to share a read only access to spend allocation page with team members as on comparison there seems to be a diff in error numbers  shown on stats page vs what is shown on spend allocation page

**Comment by mikejihbe** [comment]
FYI this feature is being reconsidered, the new relevant discussion is here -> https://github.com/getsentry/sentry/discussions/97580
